x/txfees/keeper/hooks.go

```diff
@@ -0,0 +1,62 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	epochstypes "github.com/osmosis-labs/osmosis/v7/x/epochs/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v7/x/txfees/types"
+)
+
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) {}
+
+// at the end of each epoch, swap all non-OSMO fees into OSMO and transfer to fee module account
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) {
+	nonNativeFeeAddr := k.accountKeeper.GetModuleAddress(txfeestypes.NonNativeFeeCollectorName)
+	nonNativeBalances := k.bankKeeper.GetAllBalances(ctx, nonNativeFeeAddr)
+	baseDenom, _ := k.GetBaseDenom(ctx)
+
+	for _, coin := range nonNativeBalances {
+		if coin.Denom == baseDenom {
+			continue
+		}
+		feetoken, err := k.GetFeeToken(ctx, coin.Denom)
+		if err != nil {
+			panic(err)
+		}
+
+		// We allow full slippage. Theres not really an effective way to bound slippage until TWAP's land,
+		// but even then the point is a bit moot.
+		// The only thing that could be done is a costly griefing attack to reduce the amount of osmo given as tx fees.
+		// However the idea of the txfees FeeToken gating is that the pool is sufficiently liquid for that base token.
+		_, err = k.gammKeeper.SwapExactAmountIn(ctx, nonNativeFeeAddr, feetoken.PoolID, coin, baseDenom, sdk.ZeroInt())
+		if err != nil {
+			panic(err)
+		}
```

Comment:
We should wrap this in `osmoutils.ApplyFuncIfNoErr`, and only log the error. We can't panic for an error here, as that would cause a chain halt